sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -1180,12 +1180,15 @@ object JdbcUtils extends Logging with SQLConfHelper {
     }
   }
 
-  def classifyException[T](message: String, dialect: JdbcDialect)(f: => T): T = {
+  def classifyException[T](
```

Comment:
this is a breaking change, isn't it? JDBCDialect is a public developer API.

Reply:
when was this API added? @beliefer do you know?

Reply:
https://github.com/apache/spark/pull/29952

Reply:
Perhaps we should keep two versions.

Reply:
And how will you combine exceptions from the both versions?

Reply:
How about just update the SQL migration guide and ask people to migrate on new version in 4.0.0 (the version can introduce breaking changes).

Reply:
I'm wondering if we still need this API in the jdbc dialect. We have the well-defined `FAILED_JDBC` error now, what do we expect the dialect to overwrite?

Reply:
> We have the well-defined FAILED_JDBC error now, what do we expect the dialect to overwrite?

We know which operation from Spark side failed but we don't know the root cause of the issue. A dialect can "understand" DBMS specific error better, and map some cases to Spark exceptions. Here is a simple example:
- Spark creates an index, and there is the error class `FAILED_JDBC.CREATE_INDEX` for the cases of any unclassified failures, but
- Postgres dialect knows that the SQL state `42P07` from `SQLException` belongs to existence of the object or in our case it means the index exists already. So, it can provide more precise exception `IndexAlreadyExistsException`

Reply:
there are no nested exception here. We pass a default error class to `classifyException` and if a dialect cannot classify an exception from JDBC driver, it throws the default one otherwise it forms and throws another one. The exception from JDBC driver is added as `cause`.

Reply:
So if a dialect can classify an exception, we will lose the error class which is actually worse?

Reply:
Can we also enable MiMa for this? I believe this is being skipped because it's under `execution` package.

Reply:
> So if a dialect can classify an exception, we will lose the error class which is actually worse?

@cloud-fan A dialect can override proposed error class and make our default error class more precise using the driver specific info. And yes, the original error class will be lost.

> Can we also enable MiMa for this? I believe this is being skipped because it's under execution package.

@HyukjinKwon I made the modification because MiMa complained. Did you change MiMa to expect different behaviour?

Reply:
I think losing the original error class is bad. We added `classifyException` for getting more precise java exception type, but I don't think this is useful anymore as we have error classes. My suggestion is to deprecate `classifyException`, saying that Spark never calls it anymore.